game-core/src/main/java/games/strategy/triplea/delegate/battle/FireAa.java

```diff
@@ -92,84 +89,77 @@
   public void execute(final ExecutionStack stack, final IDelegateBridge bridge) {
 
     // Loop through each type of AA and break into firing groups based on suicideOnHit
-    for (final String aaType : aaTypes) {
-      final Collection<Unit> aaTypeUnits =
-          CollectionUtils.getMatches(firingUnits, Matches.unitIsAaOfTypeAa(aaType));
-      final List<Collection<Unit>> firingGroups = MustFightBattle.newFiringUnitGroups(aaTypeUnits);
-      for (final Collection<Unit> firingGroup : firingGroups) {
-        final Set<UnitType> validTargetTypes =
-            UnitAttachment.get(firingGroup.iterator().next().getType())
-                .getTargetsAa(bridge.getData());
-        final Set<UnitType> airborneTypesTargeted =
-            defending
-                ? TechAbilityAttachment.getAirborneTargettedByAa(hitPlayer, bridge.getData())
-                    .get(aaType)
-                : new HashSet<>();
-        final Collection<Unit> validTargets =
-            CollectionUtils.getMatches(
-                attackableUnits,
-                Matches.unitIsOfTypes(validTargetTypes)
-                    .or(
-                        Matches.unitIsAirborne()
-                            .and(Matches.unitIsOfTypes(airborneTypesTargeted))));
-        final IExecutable rollDice =
-            new IExecutable() {
-              private static final long serialVersionUID = 6435935558879109347L;
+    final List<FiringGroup> groupsAndTargets =
```

Comment:
What do you think about making the `isSuicide` first dependency explicit and a condition established by the caller rather than as part of an (unenforced) contract?

The problem I can see here potentially is that it's not clear that `isSuicide` units would be first, you have to dig into the `Aa.builder()`  to know that which makes it less obvious what the code is doing from a caller perspective.

EG:
```
    // Loop through each type of AA and break into firing groups based on suicideOnHit
    final List<FiringGroup> groupsAndTargets =
        Aa.builder() //
            .aaUnits(firingUnits) //
            .aaTypes(aaTypes) //
            .hitPlayer(hitPlayer) //
            .attackableUnits(attackableUnits) //
            .defending(defending) //
            .gameData(bridge.getData()) //
            .build() //
            .getFiringGroups() //
            .stream()
            .sorted(Comparator.comparing(FiringGroup::isSuicide))
            .collect(Collectors.toList());
```
(You'll need to double check please that sorts isSuicide first, otherwise a 'reverse' or negation needs to be thrown in)

An alternative would be to document the contract of the return value:
```
Aa.builder()
   :
   :
   .getFiringGroupsWithSuicideFirst();
```

Personally I would let unit testing drive the decision. If it's easer to just unit test `Aa#getFiringGroups` without the sort, then I'd sort here. If it's easy to test, then I'd perhaps update the return method name.

Reply:
I'm not sure of the benefit of doing this.  The only caller in the current code that uses this apparently needs suicide first and the new caller in battle tree calculator would want to follow the same order.  So I'm not sure any caller would want the order to be different.

Reply:
That would be argument then for updating the name of the method to be explicit about the ordering. Otherwise it's not clear there is a contract and potentially a dependency on the ordering, essentially an undocumented behavior.

Reply:
After building unit tests, I discovered that suicide units aren't always at the front.  It first groups the firingUnits by what they can/can not target (think subs and airplanes) and then it splits those groups by isSuicide.  See RegularFiringGroupTest#getFiringGroupsWithNonTargetableUnitsAndSuicideUnits.